hip/components/prefix_sum_kernels.hip.cpp

```diff
@@ -49,21 +60,22 @@ template <typename IndexType>
 void prefix_sum(std::shared_ptr<const HipExecutor> exec, IndexType* counts,
                 size_type num_entries)
 {
-    // prefix_sum should only be performed on a valid array
-    if (num_entries > 0) {
-        auto num_blocks = ceildiv(num_entries, prefix_sum_block_size);
-        array<IndexType> block_sum_array(exec, num_blocks - 1);
-        auto block_sums = block_sum_array.get_data();
-        start_prefix_sum<prefix_sum_block_size>
-            <<<num_blocks, prefix_sum_block_size, 0, exec->get_stream()>>>(
-                num_entries, counts, block_sums);
-        // add the total sum of the previous block only when the number of
-        // blocks is larger than 1.
-        if (num_blocks > 1) {
-            finalize_prefix_sum<prefix_sum_block_size>
-                <<<num_blocks, prefix_sum_block_size, 0, exec->get_stream()>>>(
-                    num_entries, counts, block_sums);
-        }
+    constexpr auto max = std::numeric_limits<IndexType>::max();
+    array<bool> overflow{exec, {false}};
+    thrust::exclusive_scan(thrust_policy(exec), counts, counts + num_entries,
+                           counts, 0,
+                           [overflow_flag = overflow.get_data()] __device__(
+                               IndexType i, IndexType j) {
+                               auto result = i + j;
+                               if (max - i < j) {
```

Comment:
We only run prefix sums on positive arrays AFAIK, and I would keep it that way, since negative values make a great sentinel.

Reply:
from the prefix sum, it is not obvious and there's no property we can get from prefix sum with signed valued

Reply:
I checked, and we indeed only run prefix sums on nonnegative arrays. So I would suggest adding a comment to the function to make the precondition clearer.

Reply:
then I will prefer to change the name to make the nonnegative more verbose.
When using prefix sum, seldom check the documentation because prefix sum is quite commonly available in other libraries.
The additional restriction is quite unexpected

Reply:
I definitively agree with @yhmtsai . The documentation add was the bare-minimum to add this restriction, but I would also expect that the function can deal with negative numbers (as the input is signed, so negative values are permitted).
A name-change would indeed be better, or a revert back to the overflow variable (which costs us an additional allocation).

Reply:
I renamed it to prefix_sum_nonnegative, that should hopefully be unambiguous 😄